pandas/_libs/tslibs/np_datetime.pyx

```diff
@@ -678,3 +679,43 @@ cdef int64_t _convert_reso_with_dtstruct(
         raise OutOfBoundsDatetime from err
 
     return result
+
+
+@cython.overflowcheck(True)
+cpdef cnp.ndarray add_overflowsafe(cnp.ndarray left, cnp.ndarray right):
+    """
+    Overflow-safe addition for datetime64/timedelta64 dtypes.
+
+    `right` may either be zero-dim or of the same shape as `left`.
+    """
+    cdef:
+        Py_ssize_t N = left.size
+        int64_t lval, rval, res_value
+        ndarray iresult = cnp.PyArray_EMPTY(
+            left.ndim, left.shape, cnp.NPY_INT64, 0
+        )
+        cnp.broadcast mi = cnp.PyArray_MultiIterNew3(iresult, left, right)
+
+    # Note: doing this try/except outside the loop improves performance over
+    #  doing it inside the loop.
+    try:
```

Comment:
By the way I think there's another way of doing this without a loop that could also be more performant. The inspiration from this comes from a book called Hacker's Delight by Henry S Warren in section 4-2

The code would look something like:

```
uleft = left.view("uint64")
uright = right.view("uint64")
summed = uleft + uright
neg_overflow = uleft & uright & ~summed
pos_overflow = ~uleft & ~uright & summed
overflows = (neg_overflow | pos_overflow) >= 0x8000000000000000
```

The essence of the algorithm is that you use unsigned arithmetic (which has defined behavior) to do the summation, and then inspect the sign bit to see if "overflow" would have happened

Reply:
Here's a quick demo

```
In [42]: left = np.array([2 ** 63 - 1, 2 ** 63 - 1, 2 ** 63 - 1, -(2 ** 63), -(2 ** 63), -(2 ** 63)])
    ...: right = np.array([42, 1, 0, -42, -1, 0])
    ...: uleft = left.view("uint64")
    ...: uright = right.view("uint64")
    ...: summed = uleft + uright
    ...: neg_overflow = uleft & uright & ~summed
    ...: pos_overflow = ~uleft & ~uright & summed
    ...: overflows = (neg_overflow | pos_overflow) >= 0x8000000000000000
    ...: overflows
Out[42]: array([ True,  True, False,  True,  True, False])
```

Reply:
Neat.  Does this play nicely with the NPY_NAT check?

Reply:
Good callout. I suppose you would just mask that as well. So something like:

```python
In [5]: left = np.array([2 ** 63 - 1, 2 ** 63 - 1, 2 ** 63 - 1, -(2 ** 63), -(2 ** 63), -(2 ** 63)])
   ...: right = np.array([42, 1, 0, -42, -1, 0])
   ...: is_nat = (left == 2 ** 63 - 1) | (right == 2 ** 63 - 1)
   ...: uleft = left.view("uint64")
   ...: uright = right.view("uint64")
   ...: summed = uleft + uright
   ...: neg_overflow = uleft & uright & ~summed
   ...: pos_overflow = ~uleft & ~uright & summed
   ...: overflows = ((neg_overflow | pos_overflow) >= 0x8000000000000000) & ~is_nat
   ...: overflows
Out[5]: array([False, False, False,  True,  True, False])
```

I just hardcoded the above to 2 ** 63 - 1 but can replace with macro or python variable, whichever is available

Reply:
does that actually out-perform? (maybe SIMD or something?)  seems like a lot of additional memory allocations

Reply:
Hard to say. In theory yes this would leverage SIMD better if numpy uses it under the hood (not well versed in how that works). Also the overflow checks on gcc/clang are pretty fast because they use the compiler builtin, but I think MSVC falls back to a pretty slow implementation (there is a builtin for MSVC too but I don't recall Cython using it - would have to double check)

It requires more research so not something that needs to be done in this PR. Just food for future thought